### ISSUE TITLE
CORS error fetching CSS on S3 asset host while shallow routing

### ISSUE THREAD

**Issue by emrox** [vercel/next.js#21513]
**What version of Next.js are you using?**

10.0.3 & 10.0.6-canary.7

**What version of Node.js are you using?**

v14.15.1

**What browser are you using?**

Chrome, Safaro

**What operating system are you using?**

macOS, Windows, Ubuntu

**How are you deploying your application?**

next start

**Describe the Bug**

We host our app on custom EC2 instances, but our assets are deployed to S3 and served using CloudFront.
After upgrading to version `10.0.3` we noticed our app does not behave correctly by doing a full page reload instead of just rerendering the changed parts of the page. Tracking down the problem showed it looks like related to CORS headers, similar as mentioned in this bug https://github.com/vercel/next.js/issues/17029
We checked our headers, and they seem fine. 

**Expected Behavior**

page does not fully reloads when it should do shallow routing

**To Reproduce**

- create an app with a stylesheet, css or scss.
- import that stylesheet into a component as a module.
- build and upload your assets to a CDN using assetPrefix in next.config.js. Deploy your app and run in production mode.
- notice that your page fully reloads when it should do shallow routing with internally loading new data
- use Chrome or Safari (with Firefox everything seems fine)
- ensure in browser caching is enabled (when disable caching f.e. with network tab on dev tools it also works fine)

<img width="665" alt="Top 20 Eventlocations in Deutschland - Event Inc 2021-01-25 15-17-21" src="https://user-images.githubusercontent.com/62679/105717458-731ffb00-5f20-11eb-9930-d74c92e65363.png">

**Comment by emrox** [comment]
We applied this fix https://github.com/vercel/next.js/pull/18950/files, witch was mentioned on #17029 and it seems to fix that issue. (change to make it work with the current version: https://github.com/emrox/next.js/commit/f9ef0958a095d7b8fa781cbafbf50a799a0c784b)

So I'll do a PR with that one

**Comment by Manc** [comment]
I was reading through #17029 as well and I'm still not sure what needs to be done to fix this.

I upgraded from 9.4 to 10.0.5 and I'm getting the same kind of error in the browser console when the Next.js script tries to fetch (using `fetch`) a CSS file from the CDN:

> _Access to fetch at 'https://cdn.example.com/_next/static/css/f7ed95af2f1cbd99f121.css' from origin 'https://test.example.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled._

I have tried adding `crossOrigin: 'anonymous'` to `next.config.js`, but it didn't make any difference. I have not added CORS headers to the CDN responses yet, because it's quite a task and I believe this should not be necessary.

If it's not a bug (since #17029 was closed months ago), clear documentation would be very much appreciated. Thanks!

**Comment by emrox** [comment]
setting `crossOrigin` to `anonymous` or `use-credentials` also didn't solved the problem for us, that's why I've tried with that patch which helped

> clear documentation would be very much appreciated.

can you give me a hint what to document in particular?

**Comment by Manc** [comment]
I yet have to try your patch.

In terms of documentation, I guess this page is probably best suited: https://nextjs.org/docs/api-reference/next.config.js/cdn-support-with-asset-prefix

**Comment by Manc** [comment]
@emrox It took me a while now to apply the patch to my project (I used [patch-package](https://github.com/ds300/patch-package)). When I deployed the new version to my test server, there were no more errors in the console, but unfortunately when clicking on links, all HTML pages are now loaded from the server, i.e. a full page load as opposed to shallow routing.

Without the patch my website basically works fine, including shallow routing, but I get the one CORS error as described above.

In case you or anyone is interested, this is what my patch file `patches/next+10.0.5.patch` looks like:

```
diff --git a/node_modules/next/dist/client/route-loader.js b/node_modules/next/dist/client/route-loader.js
index 12bfbe4..939d32b 100644
--- a/node_modules/next/dist/client/route-loader.js
+++ b/node_modules/next/dist/client/route-loader.js
@@ -24,7 +24,19 @@ script.src=src;document.body.appendChild(script);});}function idleTimeout(ms,err
 function getClientBuildManifest(){if(self.__BUILD_MANIFEST){return Promise.resolve(self.__BUILD_MANIFEST);}const onBuildManifest=new Promise(resolve=>{// Mandatory because this is not concurrent safe:
 const cb=self.__BUILD_MANIFEST_CB;self.__BUILD_MANIFEST_CB=()=>{resolve(self.__BUILD_MANIFEST);cb&&cb();};});return Promise.race([onBuildManifest,idleTimeout(MS_MAX_IDLE_DELAY,markAssetError(new Error('Failed to load client build manifest')))]);}function getFilesForRoute(assetPrefix,route){if(process.env.NODE_ENV==='development'){return Promise.resolve({scripts:[assetPrefix+'/_next/static/chunks/pages'+encodeURI((0,_getAssetPathFromRoute.default)(route,'.js'))],// Styles are handled by `style-loader` in development:
 css:[]});}return getClientBuildManifest().then(manifest=>{if(!(route in manifest)){throw markAssetError(new Error(`Failed to lookup route: ${route}`));}const allFiles=manifest[route].map(entry=>assetPrefix+'/_next/'+encodeURI(entry));return{scripts:allFiles.filter(v=>v.endsWith('.js')),css:allFiles.filter(v=>v.endsWith('.css'))};});}function createRouteLoader(assetPrefix){const entrypoints=new Map();const loadedScripts=new Map();const styleSheets=new Map();const routes=new Map();function maybeExecuteScript(src){let prom=loadedScripts.get(src);if(prom){return prom;}// Skip executing script if it's already in the DOM:
-if(document.querySelector(`script[src^="${src}"]`)){return Promise.resolve();}loadedScripts.set(src,prom=appendScript(src));return prom;}function fetchStyleSheet(href){let prom=styleSheets.get(href);if(prom){return prom;}styleSheets.set(href,prom=fetch(href).then(res=>{if(!res.ok){throw new Error(`Failed to load stylesheet: ${href}`);}return res.text().then(text=>({href:href,content:text}));}).catch(err=>{throw markAssetError(err);}));return prom;}return{whenEntrypoint(route){return withFuture(route,entrypoints);},onEntrypoint(route,execute){Promise.resolve(execute).then(fn=>fn()).then(exports=>({component:exports&&exports.default||exports,exports:exports}),err=>({error:err})).then(input=>{const old=entrypoints.get(route);entrypoints.set(route,input);if(old&&'resolve'in old)old.resolve(input);});},loadRoute(route){return withFuture(route,routes,async()=>{try{const{scripts,css}=await getFilesForRoute(assetPrefix,route);const[,styles]=await Promise.all([entrypoints.has(route)?[]:Promise.all(scripts.map(maybeExecuteScript)),Promise.all(css.map(fetchStyleSheet))]);const entrypoint=await Promise.race([this.whenEntrypoint(route),idleTimeout(MS_MAX_IDLE_DELAY,markAssetError(new Error(`Route did not complete loading: ${route}`)))]);const res=Object.assign({styles},entrypoint);return'error'in entrypoint?entrypoint:res;}catch(err){return{error:err};}});},prefetch(route){// https://github.com/GoogleChromeLabs/quicklink/blob/453a661fa1fa940e2d2e044452398e38c67a98fb/src/index.mjs#L115-L118
+if(document.querySelector(`script[src^="${src}"]`)){return Promise.resolve();}loadedScripts.set(src,prom=appendScript(src));return prom;}function fetchStyleSheet(href){let prom=styleSheets.get(href);if(prom){return prom;}styleSheets.set(href,
+
+	prom=fetch(href,{mode:'no-cors'})
+
+.then(res=>{if(!res.ok){throw new Error(`Failed to load stylesheet: ${href}`);}return res.text().then(text=>({href:href,content:text}));}).catch(err=>{throw markAssetError(err);}));return prom;}return{whenEntrypoint(route){return withFuture(route,entrypoints);},onEntrypoint(route,execute){Promise.resolve(execute).then(fn=>fn()).then(exports=>({component:exports&&exports.default||exports,exports:exports}),err=>({error:err})).then(input=>{const old=entrypoints.get(route);entrypoints.set(route,input);if(old&&'resolve'in old)old.resolve(input);});},loadRoute(route){return withFuture(route,routes,async()=>{try{const{scripts,css}=await getFilesForRoute(assetPrefix,route);const[,styles]=await Promise.all([entrypoints.has(route)?[]:Promise.all(scripts.map(maybeExecuteScript)),
+
+Promise.all(
+	css.map(
+		(href) => document.querySelector(`link[href^="${href}"]`) || fetchStyleSheet(href)
+	)
+)
+
+]);const entrypoint=await Promise.race([this.whenEntrypoint(route),idleTimeout(MS_MAX_IDLE_DELAY,markAssetError(new Error(`Route did not complete loading: ${route}`)))]);const res=Object.assign({styles},entrypoint);return'error'in entrypoint?entrypoint:res;}catch(err){return{error:err};}});},prefetch(route){// https://github.com/GoogleChromeLabs/quicklink/blob/453a661fa1fa940e2d2e044452398e38c67a98fb/src/index.mjs#L115-L118
 // License: Apache 2.0
 let cn;if(cn=navigator.connection){// Don't prefetch if using 2G or if Save-Data is enabled.
 if(cn.saveData||/2g/.test(cn.effectiveType))return Promise.resolve();}return getFilesForRoute(assetPrefix,route).then(output=>Promise.all(canPrefetch?output.scripts.map(script=>prefetchViaDom(script,'script')):[])).then(()=>{(0,_requestIdleCallback.default)(()=>this.loadRoute(route));}).catch(// swallow prefetch errors
```

**Comment by Manc** [comment]
I've added CORS headers now to `/_next/static/*` on the CDN and everything works fine.

I don't understand why this is necessary (apparently since 9.5) or how this makes things better. Any explanation would be appreciated. I think it will also help new users and those upgrading if there was a mention of this requirement in the docs.

**Comment by emrox** [comment]
CORS is not a thing by Next.js in general, and it was there even before 9.5. Also it's not a thing implemented by Next.js since it's a general HTTP-header based mechanism.

To understand more I can recommend you these articles:
- https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
- https://dev.to/lydiahallie/cs-visualized-cors-5b8h

This problem is more specific about a problem with shallow routing we discovered.

**Comment by Manc** [comment]
Don't get me wrong, I don't want to criticise Next.js or anything (I think it's amazing). I am quite aware of CORS, but I'm pretty sure something has changed in the way Next.js loads or pre-loads resources since version 9.4. When I upgraded I had to change the CSP (specifically `style-src: 'unsafe-inline'`) and the CORS headers. This was not necessary with 9.4 and before.

I know, Next.js has an official alternative for that CSP policy (basically, shifting everything over from HTTP headers to HTML meta tags), but that is not my favourite option as it makes things more complicated, because my server pulls in HTML pages from various sources, only one of them is Next.js. I will need to look into it a bit deeper and I'm sure I will find a solution, but overall I am wondering if `fetch`-ing CSS is really the way forward and if it really has any significant benefits over the way Next.js worked before 9.5. This is totally off-topic now, so I will stop here (sorry!).

To come back to your issue – as I mentioned above, once I activated the CORS headers for my static Next.js-generated assets, the problem went away – no patch needed. My assets are also on S3 and served by CloudFront. I'm on version 10.0.5, not 10.0.3 as stated in your description.

One pitfall might be that your CORS headers are not properly set up yet while you think they are. Yesterday I tested with `curl` to check the headers of my CSS file, but for some reason I don't fully understand, cURL did show the headers I was looking for while a web browser like Chrome or Safari did *not* show those headers until I applied some more changes to the CloudFront config.

**Comment by stefan-eventinc** [comment]
We double checked the headers and they seem to be correct. Initial loading works well without errors. As I pointed out it also works well when I disable caching in Chrome & Safari. So I'm really not sure what it is, but the patch attached to that PR helped solving that issue.
I also checked other headers, like changing `Vary` to allow different origins, but that also didn't helped

**Comment by Manc** [comment]
Just to be sure, so you can double-check and compare, this is what the CSS file response, that previously caused the exact same kind of error message in the console as in the above screenshot, looks like now in my case:

![screenshot-cors](https://user-images.githubusercontent.com/811578/105962602-f3a53f80-6077-11eb-9819-4ad0c015237c.png)

Note: There are two requests for the same file. What I selected in the screenshot is the one that was triggered by `fetch`.

**Comment by emrox** [comment]
Thx for the response. I investigated again. In theory this header must not be present on normal `GET` requests, just on OPTIONS (preflight) requests (which are not shown in the Chrome network inspector)

I stumbled upon this Serverfault answer: https://serverfault.com/a/856948

TLDR; Amazon Cloudfront/S3 does not deliver the `Access-Control-*` headers on normal requests, just on `OPTIONS` requests - which is ok by definition. But also Google Chrome does not caches `OPTIONS` responses. So when a resource is already in cache and later fetched again with a XHR requests, that requests does not use the cache, but requests again with missing CORS headers. Also ok by definition. The problem is both strategies don't work together.

I'll close that issue. In case someone with Cloudfront / S3 has the same problem: that Serverfault answer has a lot of options to 'fix' that. I did it this way
- go to your Cloudfront config
- go to the `Origins and Origin Groups` tab
- edit the Origin pointing to your S3
- in `Origin Custom Headers` add header with
  - Header Name: `Origin`
  - Value: `*`
- wait a bit until Cloudfront and cache are refreshed (maybe also invalidate the cache manually)

**Comment by emrox** [comment]
nd big thank you to @Manc for the input! 🎉 🙏🏻

**Comment by fraitag** [comment]
On version 11.0.1 and 11.0.2-canary.7 I have the same problem.

**Comment by xiaolin** [comment]
For anyone that ran into this, using @emrox origin solution above and adding CORS to your s3 bucket permissions fixed the issue for me.

```json
[
    {
        "AllowedHeaders": [
            "*"
        ],
        "AllowedMethods": [
            "GET"
        ],
        "AllowedOrigins": [
            "http://yoursite.com"
        ],
        "ExposeHeaders": [],
        "MaxAgeSeconds": 3000
    },
    {
        "AllowedHeaders": [
            "*"
        ],
        "AllowedMethods": [
            "GET"
        ],
        "AllowedOrigins": [
            "https://yoursite.com"
        ],
        "ExposeHeaders": [],
        "MaxAgeSeconds": 3000
    },
]
```

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
